pkg/cache/server/server.go

```diff
@@ -34,9 +37,47 @@ func NewServer(c CompletedConfig) (*Server, error) {
 }
 
 func (s *Server) Run(ctx context.Context) error {
+	logger := klog.FromContext(ctx).WithValues("component", "cache-server")
 	server, err := s.ApiExtensions.New(genericapiserver.NewEmptyDelegate())
 	if err != nil {
 		return err
 	}
+	if err := server.GenericAPIServer.AddPostStartHook("bootstrap-cache-server", func(hookContext genericapiserver.PostStartHookContext) error {
+		logger := logger.WithValues("postStartHook", "bootstrap-cache-server")
+		if err = bootstrap.Bootstrap(goContext(hookContext), logger, s.ApiExtensionsClusterClient); err != nil {
+			logger.Error(err, "failed creating the static CustomResourcesDefinitions")
+			// nolint:nilerr
+			return nil // don't klog.Fatal. This only happens when context is cancelled.
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if err := server.GenericAPIServer.AddPostStartHook("cache-server-start-informers", func(hookContext genericapiserver.PostStartHookContext) error {
+		logger := logger.WithValues("postStartHook", "cache-server-start-informers")
+		s.ApiExtensionsSharedInformerFactory.Start(hookContext.StopCh)
+		select {
+		case <-hookContext.StopCh:
+			return nil // context closed, avoid reporting success below
+		default:
+		}
+		logger.Info("finished starting kube informers")
+		return nil
+	}); err != nil {
+		return err
+	}
 	return server.GenericAPIServer.PrepareRun().Run(ctx.Done())
 }
+
+// goContext turns the PostStartHookContext into a context.Context for use in routines that may or may not
+// run inside of a post-start-hook. The k8s APIServer wrote the post-start-hook context code before contexts
+// were part of the Go stdlib.
+func goContext(parent genericapiserver.PostStartHookContext) context.Context {
```

Comment:
We must have this as a helper somewhere - or if not, we've copy-pasted it 100 times. Could we DRY it out?

Reply:
we could, I'll try to find a good place :)